calyx/frontend/src/parser.rs

```diff
@@ -869,24 +878,133 @@ impl CalyxParser {
+                        span,
+                    }),
+                };
+                (guard, state)
+            }
+        ))
+    }
+
+    /// Parses the transition block, which comes after the `=>` in the state.
+    /// Transitions are either `Unconditional` (will always transition to a future state
+    /// no matter the assignments) or `Conditional` (with boolean expressions denoted a
+    /// `guard_state_pair` and a default)
+    fn transition(input: Node) -> ParseResult<Transition> {
+        Ok(match_nodes!(
+            input.into_children();
+            [
+                state_idx(idx)
+            ] => {
+                ast::Transition::Unconditional(idx)
+            },
+            [
+                guard_state_pair(pairs)..,
+            ] => {
+                // collects the pairs in the order the parser reads them
+                ast::Transition::Conditional(pairs.collect())
+            }
+        ))
+    }
+
+    /// Parses a single state within the `fsm` block. A state is denoted with it's
+    /// respective unsigned integer state index, which is is enumerated from `0..n`
+    /// States have assignments, which is a vector of `Wire`s, and transitions.
+    fn state(input: Node) -> ParseResult<(u64, Vec<ast::Wire>, Transition)> {
+        Ok(match_nodes!(
+            input.into_children();
+            [
+                state_idx(idx),
+                wire(wires)..,
+                transition(trans)
+            ] => {
+                let state_wires : Vec<ast::Wire> = wires.collect();
+                (idx, state_wires, trans) // we collect the state_idxs to use for sorting later
+            }
+        ))
+    }
+
+    /// Parses the `fsm` block to pull out the fsm block name, attributes, and list of rules.
+    /// A rule is the set of assignments and transitions each state within the `fsm` is assigned to.
+    fn fsm(input: Node) -> ParseResult<ast::Fsm> {
+        let span = Self::get_span(&input);
+        Ok(match_nodes!(
+            input.into_children();
+            [name_with_attribute((name, attrs)), state(states)..] => {
+                let mut state_data = Vec::new();
+
+                for (state_idx, assignments, transition) in states {
+                    let rule = ast::FSMRule {
+                        assignments,
+                        transition,
+                    };
+                    state_data.push((state_idx, rule));
+                }
+                // make sure to sort the rules by index to access the vector by state index.
+                state_data.sort_by(|(idx1, _), (idx2, _)| idx1.cmp(idx2));
+                let rules : Vec<ast::FSMRule> = state_data.into_iter().map(|(_, r)| r).collect();
+
+                ast::Fsm {
+                name,
+                attributes: attrs.add_span(span),
+                rules,
+                }
+            }
+        ))
+    }
+
+    fn connections(input: Node) -> ParseResult<ConnectionTuple> {
```

Comment:
Maybe a doc comment would be handy here (although I know it pre-existed your work).